src/template_partials/templatetags/partials.py

```diff
@@ -129,15 +125,53 @@ def _define_partial(parser, token, end_tag):
     if endpartial.contents not in acceptable_endpartials:
         parser.invalid_block_tag(endpartial, "endpartial", acceptable_endpartials)
 
-    if not hasattr(parser.origin, "partial_contents"):
-        parser.origin.partial_contents = {}
-    parser.origin.partial_contents[partial_name] = TemplateProxy(
-        nodelist, parser.origin, partial_name
-    )
+    # Store the partial nodelist in the parser.extra_data attribute, if available. (Django 5.1+)
+    # Otherwise, store it on the origin.
+    if hasattr(parser, "extra_data"):
+        parser.extra_data.setdefault("template-partials", {})
+        parser.extra_data["template-partials"][partial_name] = TemplateProxy(
+            nodelist, parser.origin, partial_name
+        )
+    else:
+        if not hasattr(parser.origin, "partial_contents"):
+            parser.origin.partial_contents = {}
+        parser.origin.partial_contents[partial_name] = TemplateProxy(
+            nodelist, parser.origin, partial_name
+        )
 
     return DefinePartialNode(partial_name, inline, nodelist)
 
 
+class SubDictionaryWrapper:
+    """
+    Wrap a parent dictionary, allowing deferred access to a sub-dictionary by key.
+
+    The parser.extra_data storage may not yet be populated when a partial node
+    is defined, so we defer access until rendering.
+    """
+
+    def __init__(self, parent_dict, lookup_key):
+        self.parent_dict = parent_dict
+        self.lookup_key = lookup_key
+
+    def __getitem__(self, key):
+        return self.parent_dict[self.lookup_key][key]
+
+
+class OriginPartialContentsWrapper:
+    """
+    Wrap parser.origin to allow deferred access to partial_contents.
+    """
+
+    def __init__(self, origin):
+        self.origin = origin
+
+    def __getitem__(self, key):
+        if not hasattr(self.origin, "partial_contents"):
+            self.origin.partial_contents = {}
```

Comment:
Why did we need this?

Reply:
Yes, we did not need it because it was defined in `_define_partial`.

Reply:
Ok, so if I followed you, you can remove it in another pr.